### PR TITLE
generic implementation for a client protocol mapper

### PR DESCRIPTION
**Pull request by BernhardBerbuir**
This is generic implementation for #14 . It supports saml and openid-connect clients. In order to support custom protocol mappers it does not enforce a specific structure.

example:
```hcl
resource "keycloak_generic_client_protocol_mapper" "Session_Duration" {
  name = "Session Duration"
  protocol = "saml"
  protocol_mapper = "saml-hardcode-attribute-mapper"
  realm_id = "${keycloak_realm.aws.id}"
  client_id = "${keycloak_saml_client.urn_amazon_webservices.id}"
  config = {
    "attribute.value" = "28800",
    "attribute.nameformat" = "Basic"
    "friendly.name" =  "Session Duration"
    "attribute.name" = "https://aws.amazon.com/SAML/Attributes/SessionDuration"
  }
} 
```

Before creating documentation and test cases it would be nice to get some feedback about the current code.

### CODE REVIEW

**Comment by mrparkers** [comment]
Before I give this a code review, I want to understand the use case for this resource.

Right now, there are specific resources for different protocol mappers (such as `keycloak_openid_user_attribute_protocol_mapper`).  I thought that implementing a separate resource for each type of protocol mapper was a better user experience because it didn't require referencing the Keycloak API docs in order to figure out the protocol mapper type (`oidc-usermodel-attribute-mapper` in this case) or the config values, which varied between different protocol mappers.

That being said, I can think of a few reasons why a generic protocol mapper resource like this would be useful:

1) If the provider doesn't support a particular protocol mapper, this resource can be used instead.
2) If you implemented a custom protocol mapper, this resource can be used to configure it, since the provider probably would never support that.

Are these the use cases you're thinking of, or is there something else you're trying to accomplish with this resource?

**Comment by BernhardBerbuir** [comment]
I completely agree with you: the generic client protocol mapper is less user friendly and more error prone. Above all, I need it for a custom protocol mapper (use case 2).  
=> If needed I can rename it to keycloak_custom_client_protocol_mapper

**Comment by mrparkers** [comment]
Perfect, sounds like we're on the same page then.  I think the code looks good so far.  I'd be happy to merge this and cut a release for you once some test cases are added.

**Comment by mrparkers** [comment]
Also, to answer your question, I think `keycloak_generic_client_protocol_mapper` is a good name for this resource.

File: docs/resources/keycloak_generic_client_protocol_mapper.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
+# keycloak_generic_client_protocol_mapper
+
+Allows for creating and managing protocol mapper for both types of clients (` and saml) within Keycloak.
```
[the remark this answers is not in the document]
Comment (BernhardBerbuir) [reply docs/resources/keycloak_generic_client_protocol_mapper.md]: I have fixed it.

**Pull request merged**